### ISSUE TITLE
High KL Divergence in GRPO with GPT2-Style Model (Due to Dropout?)

### ISSUE THREAD

**Issue by cliangX** [huggingface/trl#3500]
### ❓ Summary

I'm encountering extremely large KL divergence values (thousands to tens of thousands) when applying `GRPOTrainer` to a GPT2-style model, even at the very first training step. I suspect this is due to the heavy use of `Dropout` modules throughout the model.

I'm looking for best practices on how to stabilize KL divergence and training, especially when using architectures like GPT2 that contain extensive dropout layers.

---
### 🧠 Model Architecture

The model is a GPT2-like transformer, initialized from a supervised fine-tuned (SFT) checkpoint:
```
GPT2LMHeadModel(
  (transformer): GPT2Model(
    (wte): Embedding(50257, 1280)
    (wpe): Embedding(1024, 1280)
    (drop): Dropout(p=0.1, inplace=False)
    (h): ModuleList(
      (0-35): 36 x GPT2Block(
        (ln_1): LayerNorm((1280,), eps=1e-05, elementwise_affine=True)
        (attn): GPT2Attention(
          (c_attn): Conv1D(nf=3840, nx=1280)
          (c_proj): Conv1D(nf=1280, nx=1280)
          (attn_dropout): Dropout(p=0.1, inplace=False)
          (resid_dropout): Dropout(p=0.1, inplace=False)
        )
        (ln_2): LayerNorm((1280,), eps=1e-05, elementwise_affine=True)
        (mlp): GPT2MLP(
          (c_fc): Conv1D(nf=5120, nx=1280)
          (c_proj): Conv1D(nf=1280, nx=5120)
          (act): NewGELUActivation()
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
    (ln_f): LayerNorm((1280,), eps=1e-05, elementwise_affine=True)
  )
  (lm_head): Linear(in_features=1280, out_features=50257, bias=False)
)
```
As you can see, dropout is present in multiple submodules, including attention and MLP layers.

---

### ⚠️ Problem

When applying `GRPOTrainer`, I observe:

- Very large KL divergence values at step 0 (e.g. 3,000–20,000)
- This happens even before any policy update

After investigation, I found that the `logits` from `model.eval()` and `model.train()` differ significantly due to dropout, which causes KL to spike.

---

Since this model has already been supervised fine-tuned (SFT) on parallel data,
would it be reasonable to remove all dropout layers during GRPO training in order to stabilize KL divergence?
Or do you have a better recommendation?

### Motivation

No

### Your contribution

No

**Comment by qgallouedec** [comment]
Hey @cliangX, sorry for the very late reply, catching up on the backlog.

Your diagnosis is right: dropout causes the policy (`train()`) and reference (`eval()`) forward passes to disagree, which inflates the step-0 KL exactly as you saw. TRL already exposes a switch for this: set `disable_dropout=True` in your `GRPOConfig` and it will recursively zero out dropout in both the policy and the reference model see https://github.com/huggingface/trl/blob/da3ee53ed206ae163842f88ad757fe0e1a1808b8/trl/trainer/grpo_trainer.py#L678-L681
That's the recommended fix and is exactly equivalent to surgically removing the dropout modules.
